### DISCUSSION TITLE
Ideal number of partitions for data purge

### DISCUSSION THREAD

**Discussion by kushagk** [pgpartman/pg_partman#630] [General]
[Use case]: I need to purge Pgres data older than 30 days as a part of a compliance step. 
[Issue]: Say, if I partition my table `monthly` using timestamp, use retention as `1 month` in `partman.part_config` and use the default cron `bgw.interval `(hourly). Now, on 1st Feb, if there is a db entry with timestamp = 1st Jan, it will detach the partition of January month from main table and remove the entries made on 31st Jan. So, I end up deleting data which is one day old. 

If I partition the table daily, then my goal is achieved but the number of partitions become 30. 

[Questions]
1. What would be recommended solution here ?
2. How many partitions are a lot of partitions ? My service has a peak load of 150 Requests per minute.  Is it advisable to go for daily partitions here ?
3. What is the maximum number of partitions that are allowed in PGres. I couldn't find a value online. Few posts online do say to not go beyond 1000 partitions.

**Comment by keithf4** [comment]
As long as you're on PG version 12 or higher, you can have a quite a high number of partitions (several 1000) without noticing much of an issue.

If you want to try and keep the partition count lower, I'd say do monthly but instead of a 30 day/1 month retention do 60 days/2months. Unless there's some business requirement that no data older than 30 days can exist, there's no harm in keeping one extra partition around to ensure you keep your minimum data amount.

If you must purge older data, then you'd have to go with the daily option.

**Reply by kushagk** [reply]
Thanks for the reply. 
One correction in the issue description that I wrote: If we try running the cron on monthly basis, say on 5th Feb, partman does not delete data of 01st Jan(>30 days) since there can be a data of 31st Jan in the same partition. So on last day of Feb(28th or 29th), the bgw job will delete data from 01st Jan till 31st Jan. So, 1 month of partition and 1 month of retention can keep data up till 60 days.

Rather than keeping the partition and retention data on a daily basis, I can settle with a period of 15 days. 
1. This solves the business ask of 30 days retention 
2. It keeps the data in parent table for atleast 15 days which is mainly for debugging. 
3. Reduces the number of partitions thereby improving the query plan.

**Reply by keithf4** [reply]
I can almost guarantee you're not going to see much of a performance difference in the monthly vs daily partitioning with only keeping 30-60 days. As long as you've tested it and it retains the target amount of data you're looking for, either method should be fine.

**Reply by kushagk** [reply]
Got it. But we would want to keep the partitions to (half monthly)/fortnightly. 
Do we have a support of that: I see that the value of p_interval cannot be `fortnightly` ?

**Reply by keithf4** [reply]
Version 5 of pg_partman did away with the specially named partition schemes (daily, monthly, etc). You can now set whatever value is valid for the `interval` data type. So even if you're on version 4.x of partman, I would plan around that.

If you have a business requirement that you cannot have data older than 30 days in the table, stick with daily partitioning.

If you're just trying to keep the number of partitions to a minimum, I would recommend monthly and setting the retention to the smallest amount to keep the data you need.

You could go with a `2 week` interval or `14 days`, but I think that's just kind of over complicating things for next to no benefit.

**Reply by kushagk** [reply]
Understood. 
Also, how can the order of detaching/deleting partitions be controlled given one table has a dependency on other(via foreign keys)?

**Reply by kushagk** [reply]
@keithf4 : Were you able to check my comment ?
1. I have to delete partitions of 3 tables where one is linked to the next via a foreign key. 
2. Also, what are the allowed values of p_retention ? I cannot find that in the documentation. Is `1 hours` (plural) allowed or `1 hour` ?

**Reply by keithf4** [reply]
Foreign keys TO the partition set are very very tricky. Even if you do it in the right order, it may still not allow it because you cannot do a DROP CASCADE with partitioned tables without some consequences. If you do that, it drops the entire foreign key for the entire partition set, not just the FK for the child table. So be very careful. partman does not drop child tables with CASCADE, so if you run into this, maintenance will just fail and throw an error that the table cannot be dropped. If you hit that, you'll have to clear out all the referenced data first before you can drop the child table.  This is a limitation of PG itself, not something with partman.

The order that partition sets are maintained cannot be controlled at this time, but partman version 5.1 will allow users to set the order

p_retention can be any valid value for the INTERVAL data type in Postgres for time-based partitioning

**Reply by kushagk** [reply]
How about this approach: (Table1 --(FK1)--> Table2 --(FK2)--> Table3) 
1. partition the table with equal duration. 
2. Set retention value of T3 as 28 days, T2 as 29 days, T1 as 30 days. So, T3 will be deleted first, T2 next and when T1 is dropped both T2 and T3's partitions would be gone.

**Reply by keithf4** [reply]
You'd have to try it and see :)

**Reply by kushagk** [reply]
@keithf4 : Yes, I tried and it did not work :(

This is what I am going for next:
1. keep retention_keep_table = true: So the older data will not come up while querying the parent table.
2. Now, the detached partitions are free from the foreign key's barriers and can be dropped. 
3. Write a procedure to delete the detached partitioned tables since the table name follows a timestamp regular expression. 
Something like: https://stackoverflow.com/questions/13110225/drop-tables-using-regex-mysql#:~:text=The%20revised%20solution%20would%20be,GROUP_CONCAT(table_name))%20FROM%20information_schema.

Please opine.

**Reply by keithf4** [reply]
It may not even let you detach them depending on the FK data

**Reply by kushagk** [reply]
@keithf4 

It does; I tested it. 
1. I detached the the table(retention_keep_table = true) in the order of the foreign keys. The lower most table in the hierarchy had the least retention duration. 
2. Post detaching, it allows to drop the detached table if done in the right order. [This as of now is still a manual step for me and I am looking for the best ways].
3. Did not ask to break any FK relations amongst the parent table. 

This behaviour is very special compared to below:
In case of drop (retention_keep_table = false), the partitions are not dropped even if there is no data in it.
